### ISSUE TITLE
When trying to use Localized Account Execution w/o prevent taskbar grouping disabled, Launcher_LinkedProcess.Attach() fails as the intended process is not set to GW2 (or set at all)

### ISSUE THREAD

**Issue by ShiftLeftLogical** [Healix/Gw2Launcher#18]
First Issue: If you take a look at the process 'p' it is never set such that gw2-64 as the process, it is just the base process you get back from System when calling new, and since this process is not started, it's Process.Id and the like member variables are not valid:

Exception thrown: 'System.Threading.Tasks.TaskCanceledException' in mscorlib.dll
The thread 0x5768 has exited with code 0 (0x0).
Exception thrown: 'System.InvalidOperationException' in System.dll
[4/27/2019 7:37:32 PM][LinkedProcess._Attach]: No process is associated with this object.
----------------------------->
   at System.Diagnostics.Process.EnsureState(State state)
   at System.Diagnostics.Process.EnsureState(State state)
   at System.Diagnostics.Process.get_Id()
   at Gw2Launcher.Client.Launcher.LinkedProcess._Attach(Process p) in E:\GitHub\Gw2Launcher\Gw2Launcher\Client\Launcher_LinkedProcess.cs:line 167

I noticed the if prevent taskbar grouping logic so I enabled that, which did correctly grab the GW2 process but then we fail when trying to call ensure state on a process we did not start or no longer own (not a windows person here, so I don't know if a fork is being done but either way we are not the parent):

Exception thrown: 'System.InvalidOperationException' in System.dll
[4/27/2019 7:33:38 PM][Launcher.OnScanComplete]: Process was not started by this object, so requested information cannot be determined.
----------------------------->
   at System.Diagnostics.Process.EnsureState(State state)
   at System.Diagnostics.Process.get_ExitTime()
   at Gw2Launcher.Client.Launcher.OnScanComplete(DateTime startTime, ScanOptions options) in E:\GitHub\Gw2Launcher\Gw2Launcher\Client\Launcher.cs:line 2965

The reason I am going this route is that I cannot run three accounts at the same time w/o one of them, typically crashing/locking up due to what appears to be CoherentUI having shared resource/lock issues (not designed for it), it looks like the "normal" way of running Gw2Launcher still depends on one main CoherentUI as my first account uses the CoherentUI located next to the GW2 exe, not its own instanced version as do subsequent accounts. This behavior is required/intended, correct?

**Comment by Healix** [comment]
Is there a problem with the exceptions? Both of them are expected behavior and should only be seen if you're running the debugger. The first one is a method used for both new or existing processes, hence 'p' not being set when the process is first being started. The other one isn't important.

As for CoherentUI, you can find a fix to this problem under Settings > General > Launching > Localized account execution, which forces each account to use its own bin folder. This problem is caused by how GW2 wants an exclusive lock on the files and how for a brief moment it releases that lock, allowing another account to steal it, which results in GW2 crashing with a CoherentUI error.

> it looks like the "normal" way of running Gw2Launcher still depends on one main CoherentUI as my first account uses the CoherentUI located next to the GW2 exe, not its own instanced version as do subsequent accounts. This behavior is required/intended, correct?

The taskbar grouping option has no impact on how CoherentUI is used. GW2 will launch CoherentUI from the bin folder as long as those files aren't locked, otherwise it'll copy them to account's gw2cache folder. This problem is mainly encountered due to how quickly Gw2Launcher can launch multiple accounts, which was originally limited to once every 2 seconds.

**Comment by ShiftLeftLogical** [comment]
I'll need to run release and see what is going on as gw2launcher's launched gw2 instances were quickly killed/crashing or not being started as I did try and go to use the localization / exe/dll isolation route and the effect appeared to be the same as the end result of using the debugger. I'll update when I figure out what the issue is.

**Comment by Healix** [comment]
You can run it in debug mode if you want, just note that the exceptions being logged are generally not important, and the ones that are will result in a failed launch, where "failed" will be shown next to "Last used" on the account's button and hovering over it will show the reason.

If CoherentUI is the main problem, note that GW2 would usually crash with CoherentUI referenced at the top of its crash log.

For the localized execution option, note that a Gw2Launcher folder is created in your GW2 folder and Gw2-64.exe is launched from there. If you had any software that needed to be manually configured for Gw2-64.exe, you'll need to do the same here. Some people need to manually configure Gw2-64.exe to use their desired GPU for example.

If you're attempting to launch multiple accounts in either full screen or borderless, note that GW2 can fail to initialize if it loses focus while loading the character select, which will cause it to simply close. GW2 may also freeze during loading, especially when launching multiple accounts at the same time, resulting in a white or black screen for up to a minute or more depending on how stressed your system is (watch the CPU / disk usage in the task manager).

**Comment by ShiftLeftLogical** [comment]
Thanks for the reply.

I got another chance to look into this, and figured my way around msvs 2019 ... how to get debugging symbols for release build per the project file's advanced config ... etc., essentially not pounddefing DEBUG a la the debug build.

Still ran into the same issue where the gw launcher will spin up and the be quickly closed ONLY in the case that I do have "Localized account execution" enabled. I even went to the extent of installing your prebuilt 64 bit client and starting completely from scratch for my .dat and gfx. If the localized option is enabled, in this case (all cases, for me) it asks to run in normal mode (so I can edit associated files ... yet this should be editing non-shared files correct?) per usual when setting up an account, it downloads the basic data and the launcher crashes, no entering my credentials, just closed and the account will show as "Last Used" never, forever. If I disable localized account execution, all goes as normal.

Hence debugging the issue in the first place.

I noticed with release and even creating a new arch build target for just x64 not "any cpu" (same results as any cpu, in the end) I eventually get an exception at some point when trying to attach to the process which looks like a 32 bit process/lib/api trying to be invoked. It is a general system invalid operation exception with p.MainModule exception 'p.MainModule' threw an exception of type 'System.ComponentModel.Win32Exception', which from what I am reading is not something you can straight up do with a 64 bit binary (access pid 4, MainModule).

I am guessing you are going to want a call stack as I assume this does not happen to you personally? Anything else? Have you attempted, in a virtual machine, a "fresh" configuration/install of windows and gw2 as would some random Gw2Launcher user to rule out any configuration details your system might have?

**Comment by Healix** [comment]
When you say the launcher closes, how does it close? Does the launcher show at all? Does it just exit, or does it show the error reporting window, in which case you can click show details to see the log (if CoherentUI is crashing, the last part of the log is where the log for that is shown, under the error logs section).

Gw2Launcher will ask if you want to launch in normal mode whenever you're using a Local.dat file for the first time. This is because GW2 cannot modify any of its files when launching in shared mode, so a new file must first either be updated or launched in normal mode. The primary reason for this message is to make it known (again) that launching in normal mode is required to make changes to Local.dat, which is where GW2 stores your email/password and other hardware-related settings.

"Last used" only updates when the game launches fully to character select.

"p.MainModule" can't be used between 32-bit and 64-bit processes. I never fully rely on using this and any errors it causes should be handled. Note that the exceptions I print in the log are just for reference and most of them don't mean anything.

When using the localized execution option, note that the only major difference is that your path to Gw2-64.exe/Gw2.exe changes. If you have any software that needed to be configured for Gw2-64.exe, notably anti-virus/security software, you'll need to configure it for this new path. After trying to launch an account with this option enabled, check your GW2 folder (where Gw2-64.exe is located) for a Gw2Launcher folder. Within this folder should be numbered folders, one for each account you've launched (if you want to know the ID of an account, right click an account > edit > statistics). In these numbered folders are replicas of your GW2 folder (note that these files are fake and are basically shortcuts to the real ones). You can try launching Gw2-64.exe directly from one of these folders to confirm it's actually working.

For testing, I use Windows 10, 7 and 10 (clean) in a VM.

**Comment by ShiftLeftLogical** [comment]
Without localization the launcher shows, quickly then WerFault.exe came up which tipped me off that I am sending crash data off to MS.

Here is the application error but it does not tell me much past the c runtime reporting a stack overflow:

> Faulting application name: Gw2-64.exe, version: 1.0.0.1, time stamp: 0x5cc2351c
> Faulting module name: ucrtbase.dll, version: 10.0.17134.677, time stamp: 0x9f346d3f
> Exception code: **0xc0000409**
> Fault offset: 0x000000000006ca78
> Faulting process id: 0x3a8
> Faulting application start time: 0x01d506e2da39ac87
> Faulting application path: F:\Program Files\Guild Wars 2\Gw2Launcher\1\Gw2-64.exe
> Faulting module path: C:\Windows\System32\ucrtbase.dll
> Report Id: 171741e9-fd3a-4710-85b1-7b27feedface
> Faulting package full name: 
> Faulting package-relative application ID:

**Comment by Healix** [comment]
> Without localization the launcher shows quickly, then WerFault.exe came up which tipped me off that I am sending crash data off to MS.

It's crashing **without** the localization option? I'm assuming you meant **with** the option enabled, based on the crash report.

Try running this Gw2-64.exe directly (F:\Program Files\Guild Wars 2\Gw2Launcher\1\Gw2-64.exe) to see if it still crashes.

**Comment by ShiftLeftLogical** [comment]
Yes, sorry, with localization, this is where I have the issue.

It does not crash w/o localization ... works just fine in terms of launching the game clients.

**Comment by ShiftLeftLogical** [comment]
Same issue:

Faulting application name: Gw2-64.exe, version: 1.0.0.1, time stamp: 0x5cc2351c
Faulting module name: ucrtbase.dll, version: 10.0.17134.677, time stamp: 0x9f346d3f
Exception code: 0xc0000409
Fault offset: 0x000000000006ca78
Faulting process id: 0x3418
Faulting application start time: 0x01d507b2c2b45d4b
**Faulting application path: F:\Program Files\Guild Wars 2\Gw2Launcher\1\Gw2-64.exe**
Faulting module path: C:\Windows\System32\ucrtbase.dll
Report Id: 2bea8074-9f05-45b8-8142-1507feedface
Faulting package full name: 
Faulting package-relative application ID: 

Note the same path to the .exe as when launched via the gw2launcher GUI, the path for localization is correct, just that the "localized" copy has this above crash.

**Comment by ShiftLeftLogical** [comment]
Figured it out.

I am using, d912pxy and obviously you have no context as to what thirdparty additions I am using and in turn no reason to know I needed to have the d912pxy folder adjacent to the bin64 directory (F:\Program Files\Guild Wars 2\Gw2Launcher\1) which contains shader caches and pre-compiled shaders used by the modified d3d9.dll which you are copying over when enabling localization.

Quite the stupid move on my part thinking the modified d3d9.dll copy was all that was needed when it's modifications rely on shader caches at a relative offset to the Gw2-64.exe and in turn, bin64 folder.

I apologize for dragging you through that. I will fork your code and just add the simple copy in the case that d912pxy exists, by virtue of the folder being in that location.

Now, to see how d912pxy behaves when running multiboxed. I know the d912pxy dev still has mutex issues to solve for multiboxing and is why I was looking for this type of solution you provided. I also was having CoherentUI crashes due to multiboxing (rarer) so I will see how this solution behaves. At this point I will close  and any issues encountered, I will restore the d3d9.dll original prior to filing against this project.

Thank you for your work on this, not to mention the clean code ... made it very easy to follow for a non-windows programmer.
